src/dbnode/client/session.go

```diff
@@ -3143,10 +3151,12 @@ func (b *baseBlocksResult) segmentForBlock(seg *rpc.Segment) ts.Segment {
 
 func (b *baseBlocksResult) mergeReaders(start time.Time, blockSize time.Duration, readers []xio.SegmentReader) (encoding.Encoder, error) {
 	iter := b.multiReaderIteratorPool.Get()
+	iter.SetSchema(b.nsCtx.Schema)
 	iter.Reset(readers, start, blockSize)
 	defer iter.Close()
 
 	encoder := b.encoderPool.Get()
+	encoder.SetSchema(b.nsCtx.Schema)
```

Comment:
Similarly I'd be in favor of making `schema` be required by call to encoder `Reset(...)` rather than being a separate method since it's required every time you start using the iterator anyhow.

Reply:
encoder is slightly different. we could call SetSchema in the middle.
Still, I would make schema part of reset, but keep a separate SetSchema for encoder to allow mid stream schema change.